test/flush_worker_test.cc

```diff
@@ -0,0 +1,149 @@
+    // mock_api_ takes the ownership of dispatcher_.
+    EXPECT_CALL(mock_api_, allocateDispatcher_(_, _)).WillOnce(Return(dispatcher_));
+  }
+
+  // We set up emulating timer firing and moving simulated time forward in
+  // simulateTimerloop() below.
+  void setupDispatcherTimerEmulation() {
+    timer_ = new NiceMock<Envoy::Event::MockTimer>();
+    EXPECT_CALL(*dispatcher_, createTimer_(_)).WillOnce(Invoke([&](Envoy::Event::TimerCb cb) {
+      timer_cb_ = std::move(cb);
+      return timer_;
+    }));
+    EXPECT_CALL(*timer_, enableTimer(_, _))
+        .WillRepeatedly(Invoke([&](const std::chrono::microseconds,
+                                   const Envoy::ScopeTrackedObject*) { timer_set_ = true; }));
+    EXPECT_CALL(*timer_, disableTimer()).WillOnce(Invoke([&]() { timer_set_ = false; }));
+    EXPECT_CALL(*dispatcher_, exit()).WillOnce(Invoke([&]() {
+      auto guard = std::make_unique<Envoy::Thread::LockGuard>(lock_);
+      stopped_ = true;
+    }));
+  }
+
+  void expectDispatcherRun() {
+    EXPECT_CALL(*dispatcher_, run(_))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The first dispatcher run is in WorkerImpl::start().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::NonBlock, type);
+        }))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The second dispatcher run is in FlushWorkerImpl::work().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::RunUntilExit, type);
+          simulateTimerLoop();
+        }));
+  }
+
+  void simulateTimerLoop() {
+    int i = 0;
+    while (true) {
+      i++;
+      // At least run the while loop for 100 times.
+      if (i == 100) {
+        promise_.set_value();
+      }
+      auto guard = std::make_unique<Envoy::Thread::LockGuard>(lock_);
+      if (stopped_) {
+        break;
+      } else if (timer_set_) {
+        timer_set_ = false;
+        timer_cb_();
+      }
+    }
+  }
+
+  NiceMock<Envoy::Api::MockApi> mock_api_;
+  Envoy::Thread::ThreadFactory& thread_factory_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  NiceMock<Envoy::ThreadLocal::MockInstance> tls_;
+  Envoy::Random::RandomGeneratorImpl rand_;
+  NiceMock<Envoy::Event::MockDispatcher>* dispatcher_ = nullptr; // not owned
+  std::unique_ptr<Envoy::Runtime::ScopedLoaderSingleton> loader_;
+  NiceMock<Envoy::LocalInfo::MockLocalInfo> local_info_;
+  NiceMock<Envoy::ProtobufMessage::MockValidationVisitor> validation_visitor_;
+
+  NiceMock<Envoy::Event::MockTimer>* timer_; // not owned
+  Envoy::Event::TimerCb timer_cb_;
+  bool timer_set_{};
+  bool stopped_{false};
+  // Protect stopped_.
+  Envoy::Thread::MutexBasicLockable lock_;
+  std::promise<void> promise_;
+
+  Envoy::Stats::MockSink* sink_ = nullptr; // owned by stats_sinks_
+  std::list<std::unique_ptr<Envoy::Stats::Sink>> stats_sinks_;
+  std::chrono::milliseconds stats_flush_interval_{10};
```

Comment:
This test seems really complicated, and is very hard to read.  I'm really having trouble following it.  One of the problems is you've declared 18 member variables.  Is that actually necessary here?  Can some of those be moved to local variables?  Local variables have the advantage of telling the user what they are when they are used.

Some more comments all around might make this easier as well.

Also, better naming will help.  Calling your promise promise_ tells us almost nothing.  What is being promised?

Reply:
Thanks Nate. Changed rand_, local_info_, validation_visitor_, stats_flush_interval_ to local variables. The rest member variables are needed.

Also added some comments and renamed promise_ to signal_dispatcher_to_exit_ since it is used to guarantee `dispatcher->exit()` is called after `signal_dispatcher_to_exit_.set_value();` (the while loop runs at least 100 times).

Yes, agree that the test is difficult to read. Partly it is because we are trying to simulate that a periodical timer and a callback keeps firing until a stop signal from another thread.